lib/Filesystem/Domain/FileList.php

```diff
@@ -86,30 +86,50 @@ public function byExtensions(array $extensions): self
     }
 
     /**
-     * @param string[] $globPatterns
+     * @param string[] $includePatterns
+     * @param string[] $excludePatterns
      */
-    public function excludePatterns(array $globPatterns): self
+    public function includeAndExclude(array $includePatterns = [], array $excludePatterns = []): self
     {
-        return $this->filter(function (SplFileInfo $info) use ($globPatterns) {
-            foreach ($globPatterns as $pattern) {
-                if (Glob::match($info->getPathname(), $pattern)) {
-                    return false;
-                }
+        $inclusionMap = [];
+        if ($includePatterns === []) {
+            $inclusionMap['/**/*'] = true;
+        }
+
+        foreach ($includePatterns as $includePattern) {
+            $inclusionMap[$includePattern] = true;
+        }
+        foreach ($excludePatterns as $excludePattern) {
+            $inclusionMap[$excludePattern] = false;
+        }
+
+        // Sort map by keys so that more specific paths are getting matched first
+        uksort($inclusionMap, function (string $x, string $y) {
+            $partsX = explode(DIRECTORY_SEPARATOR, $x);
+            $partsY = explode(DIRECTORY_SEPARATOR, $y);
+            // Longer paths should come first
+            $countDiff = -(count($partsX) <=> count($partsY));
```

Comment:
maybe remove the sign here and the brackets and return `-$countDiff`

Reply:
Done